metrics/src/control.rs

```diff
@@ -46,16 +53,39 @@ impl Controller {
             .map_err(|_| SnapshotError::ReceiverShutdown)
             .and_then(move |_| rx.recv().map_err(|_| SnapshotError::InternalError))
     }
+}
+
+impl AsyncSnapshotProvider for Controller {
+    type Snapshot = Snapshot;
+    type SnapshotError = SnapshotError;
+    type SnapshotFuture = SnapshotFuture;
 
-    /// Retrieves a snapshot of the current metric state asynchronously.
-    pub fn get_snapshot_async(&self) -> Result<oneshot::Receiver<Snapshot>, SnapshotError> {
+    /// Gets a snapshot asynchronously.
+    fn get_snapshot_async(&self) -> Self::SnapshotFuture {
         let (tx, rx) = oneshot::channel();
         let msg = ControlFrame::SnapshotAsync(tx);
 
         self.control_tx
             .send(msg)
-            .map_err(|_| SnapshotError::ReceiverShutdown)
-            .map(move |_| rx)
+            .map(move |_| SnapshotFuture::Waiting(rx))
+            .unwrap_or(SnapshotFuture::Errored(SnapshotError::ReceiverShutdown))
+    }
+}
+
+pub enum SnapshotFuture {
+    Waiting(oneshot::Receiver<Snapshot>),
+    Errored(SnapshotError),
```

Comment:
not totally sure you need this variant? Can't you just poll the future and check its internal error?

Reply:
It handles the case when there's an error even sending the command to the receiver i.e. the receiver thread is dead so nobody can receive our request to generate a snapshot.

Thinking about it, though... maybe if the sender of the oneshot pair we generate drops because it was never able to be sent, then the receiver of the pair would implicitly error as the sender dropped without sending a value... so we could likely drop this variant and switch to a pure struct.  I'll play around with that.

Reply:
Thinking through this a little more: I think I want to leave this as is.

Essentially, two errors can happen:
- the receiver is shut down (thread is stopped, thread panics, whatever)
- an error occurs specifically during generating (and sending back) the snapshot

If we rely on the `oneshot::Receiver` to give us back an error, the only error, we wouldn't know if the receiver was shut down or if an error occurred during the generation of the snapshot.  I'd like to keep it as is because it lets us inform the user if the receiver is shutdown, which, more likely than not, will primarily be when they forget to start it.

It's just a nice little differentiation, IMO, to give a more useful error message between the two cases without any performance loss or loss of ergonomics.